src/Nethermind/Nethermind.Facade/Eth/RpcTransaction/TransactionForRpc.cs

```diff
@@ -123,27 +123,29 @@ internal static void RegisterTransactionType<T>() where T : TransactionForRpc, I
 
         public override TransactionForRpc? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
-            const string TypeFieldKey = nameof(TransactionForRpc.Type);
             // Copy the reader so we can do a double parse:
             // The first parse is used to check for fields, while the second parses the entire Transaction
             Utf8JsonReader txTypeReader = reader;
             JsonObject untyped = JsonSerializer.Deserialize<JsonObject>(ref txTypeReader, options);
 
-            TxType? txType = null;
-            if (untyped.TryGetPropertyValue(TypeFieldKey, out JsonNode? node))
+            Type concreteTxType = DeriveTxType(untyped, options);
+
+            return (TransactionForRpc?)JsonSerializer.Deserialize(ref reader, concreteTxType, options);
+        }
+
+        private Type DeriveTxType(JsonObject untyped, JsonSerializerOptions options)
+        {
+            const TxType defaultTxType = TxType.EIP1559;
+            const string gasPriceFieldKey = nameof(LegacyTransactionForRpc.GasPrice);
+
+            if (untyped.ContainsKey(gasPriceFieldKey))
             {
-                txType = node.Deserialize<TxType?>(options);
+                return typeof(LegacyTransactionForRpc);
             }
 
-            Type concreteTxType =
-                (
-                    txType is not null
-                        ? _txTypes.FirstOrDefault(p => p.TxType == txType)
-                        : _txTypes.FirstOrDefault(p => p.DiscriminatorProperties.Any(name => untyped.ContainsKey(name)), _txTypes[^1])
-                )?.Type
-                ?? throw new JsonException("Unknown transaction type");
-
-            return (TransactionForRpc?)JsonSerializer.Deserialize(ref reader, concreteTxType, options);
+            return _txTypes
+                .FirstOrDefault(p => p.TxType == defaultTxType || p.DiscriminatorProperties.Any(untyped.ContainsKey),
+                    _txTypes[^1]).Type;
```

Comment:
I don't get it...

Won't this often default to `defaultTxType` and it is based on order?

I would expect something like below:

```suggestion
            return _txTypes.FirstOrDefault(p => p.DiscriminatorProperties.Any(untyped.ContainsKey))?.Type ?? defaultTxType;
```